storage/allocator.go

```diff
@@ -155,46 +155,59 @@ func (a *allocator) capacityGossipUpdate(key string, _ bool) {
 	a.capacityKeys[key] = struct{}{}
 }
 
-// AllocateTarget returns a suitable store for a new allocation with the
-// required attributes. Nodes already accommodating existing replicas
-// are ruled out as targets.
-func (a *allocator) AllocateTarget(required proto.Attributes, existing []proto.Replica) (
-	*proto.StoreDescriptor, error) {
+// AllocateTarget returns a suitable store for a new allocation with
+// the required attributes. Nodes already accommodating existing
+// replicas are ruled out as targets. If relaxConstraints is true,
+// then the required attributes will be relaxed as necessary, from
+// least specific to most specific, in order to allocate a target.
+func (a *allocator) AllocateTarget(required proto.Attributes, existing []proto.Replica,
+	relaxConstraints bool) (*proto.StoreDescriptor, error) {
 	a.Lock()
 	defer a.Unlock()
-	return a.allocateTargetInternal(required, existing, nil)
+	return a.allocateTargetInternal(required, existing, relaxConstraints, nil)
 }
 
 func (a *allocator) allocateTargetInternal(required proto.Attributes, existing []proto.Replica,
-	filter func(*proto.StoreDescriptor, *stat, *stat) bool) (*proto.StoreDescriptor, error) {
-	stores, sl, err := a.selectRandom(3, required, existing)
-	if err != nil {
-		return nil, err
-	}
+	relaxConstraints bool, filter func(*proto.StoreDescriptor, *stat, *stat) bool) (*proto.StoreDescriptor, error) {
+	attrs := append([]string(nil), required.Attrs...)
```

Comment:
I really don't know if it's faster. You could imagine that it doesn't do an allocation and []string{} does, though I doubt go is that stupid.

Not worried about the copy.